File: service/src/main/java/hu/bsstudio/scores/audience/DefaultAudienceScoreService.java
package hu.bsstudio.scores.audience;

import hu.bsstudio.entity.TeamEntity;
import hu.bsstudio.repository.TeamRepository;
import hu.bsstudio.scores.audience.model.AudienceScoredTeam;
import hu.bsstudio.team.mapper.TeamModelEntityMapper;
import hu.bsstudio.team.model.DetailedTeam;
import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import reactor.core.publisher.Mono;

@RequiredArgsConstructor
public class DefaultAudienceScoreService implements AudienceScoreService {

    private final TeamRepository teamRepository;

    @Setter(AccessLevel.PACKAGE)
    private TeamModelEntityMapper teamModelEntityMapper = new TeamModelEntityMapper();

    @Override
    public Mono<DetailedTeam> updateAudienceScore(final AudienceScoredTeam audienceScoredTeam) {
        return Mono.just(audienceScoredTeam)
            .map(AudienceScoredTeam::getTeamId)
            .flatMap(teamRepository::findById)
            .map(entity -> addAudienceScore(entity, audienceScoredTeam))
            .flatMap(teamRepository::save)
            .map(teamModelEntityMapper::toModel);
    }

    private TeamEntity addAudienceScore(final TeamEntity entity, final AudienceScoredTeam audienceScoredTeam) {
        entity.setVotes(audienceScoredTeam.getVotes());
        entity.setAudienceScore(audienceScoredTeam.getAudienceScore());
        return entity;
    }
}

File: service/src/main/kotlin/hu/bsstudio/robonaut/scores/audience/DefaultAudienceScoreService.kt
package hu.bsstudio.robonaut.scores.audience

import hu.bsstudio.robonaut.entity.TeamEntity
import hu.bsstudio.robonaut.repository.TeamRepository
import hu.bsstudio.robonaut.scores.audience.model.AudienceScoredTeam
import hu.bsstudio.robonaut.team.mapper.TeamModelEntityMapper
import hu.bsstudio.robonaut.team.model.DetailedTeam
import reactor.core.publisher.Mono

class DefaultAudienceScoreService(
    private val teamRepository: TeamRepository,
    private val teamModelEntityMapper: TeamModelEntityMapper = TeamModelEntityMapper(),
) : AudienceScoreService {
    override fun updateAudienceScore(audienceScoredTeam: AudienceScoredTeam): Mono<DetailedTeam> {
        return Mono.just(audienceScoredTeam)
            .map(AudienceScoredTeam::teamId)
            .flatMap(teamRepository::findById)
            .map { addAudienceScore(it, audienceScoredTeam) }
            .flatMap(teamRepository::save)
            .map(teamModelEntityMapper::toModel)
    }

    private fun addAudienceScore(
        entity: TeamEntity,
        audienceScoredTeam: AudienceScoredTeam,
    ): TeamEntity {
        entity.votes = audienceScoredTeam.votes
        entity.audienceScore = audienceScoredTeam.audienceScore
        return entity
    }
}
